num/quat/exp.go

```diff
@@ -0,0 +1,64 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package quat
+
+import "math"
+
+// Exp returns e**q, the base-e exponential of q.
+func Exp(q Quat) Quat {
+	w, uv := split(q)
+	if uv == zero {
+		return lift(math.Exp(w))
+	}
+	v := Abs(uv)
+	e := math.Exp(w)
+	s, c := math.Sincos(v)
+	return Add(lift(e*c), Scale(e*s/v, uv))
+}
+
+// Log returns the natural logarithm of q.
+func Log(q Quat) Quat {
+	w, uv := split(q)
+	if uv == zero {
+		return lift(math.Log(w))
+	}
+	v := Abs(uv)
+	return Add(lift(math.Log(Abs(q))), Scale(math.Atan2(v, w)/v, uv))
+}
+
+// Pow return q**r, the base-q exponential of r.
+// For generalized compatibility with math.Pow:
+//      Pow(0, ±0) returns 1+0i+0j+0k
+//      Pow(0, c) for real(c)<0 returns Inf+0i+0j+0k if imag(c), jmag(c), kmag(c) are zero,
+//          otherwise Inf+Inf i+Inf j+Inf k.
+func Pow(q, r Quat) Quat {
+	if q == zero {
+		w, uv := split(r)
```

Comment:
I don't think this switch statement does what the documentation says. The documentation is about the power, but the switch tests the real part of the quaternion.

Reply:
I don't understand; the switch tests the real part of the quaternion that q is raised to. There is some leeway; the doc says Pow(0, ±0) returns 1+0i+0j+0k, but the switch only tests the real part. I don't have enough knowledge of quaternion math to knwo whether this is OK.

Reply:
The documentation regarding `±0` here is unclear and even misleading but since the implementation follows `cmplx.Pow`, I think I'd leave it as is.